### PR TITLE
JDBC enable usage of LocalDateTime via get/setObject

### PR DESCRIPTION
**Pull request by Jens-H**
This PR implements the methods to use getObject and setObject with java.time.LocalDateTime. E.g.

`getObject(1, java.time.LocalDateTime.class)`

Please note that the getObject implementation is not complete for all possible data types and combinations. However, the goal to use LocalDateTime is completely achieved.

Also some test are added to the TestDuckDBJDBC class.

~~As PR #2888 is not merged yet, this one includes both changes.~~

### CODE REVIEW

File: tools/jdbc/src/test/java/org/duckdb/test/TestDuckDBJDBC.java [hunk 1]
```diff
@@ -410,7 +411,63 @@ public static void duckdb_timestamp_test() throws Exception {
 
 		conn.close();
 	}
+
+	public static void test_duckdb_localdatetime() throws Exception {
+		Connection conn = DriverManager.getConnection("jdbc:duckdb:");
+		Statement stmt = conn.createStatement();
+		stmt.execute("CREATE TABLE x (ts TIMESTAMP)");
+
+		LocalDateTime ldt = LocalDateTime.parse("2021-01-18T21:20:07");
```
Comment (hawkfish) [inline tools/jdbc/src/test/java/org/duckdb/test/TestDuckDBJDBC.java:420]: Will the instant specified by the string vary by testing time zone? Does it need an offset?

File: tools/jdbc/src/test/java/org/duckdb/test/TestDuckDBJDBC.java [hunk 2]
```diff
@@ -410,7 +411,63 @@ public static void duckdb_timestamp_test() throws Exception {
...
+			+ " flt FLOAT, dbl DOUBLE, dte DATE, tme TIME, ts TIMESTAMP, dec16 DECIMAL(3,1),"
+			+ " dec32 DECIMAL(9,8), dec64 DECIMAL(16,1), dec128 DECIMAL(30,10))");
+		stmt.execute("INSERT INTO b VALUES ('varchary', true, 6, 42, 666, 42.666, 666.42,"
+			+ " '1970-01-02', '01:00:34', '1970-01-03 03:42:23', 42.2, 1.23456789, 987654321012345.6, 111112222233333.44444)");
+
+		PreparedStatement ps = conn.prepareStatement("SELECT * FROM b");
+		ResultSet rs = ps.executeQuery();
+
+		rs.next();
+		assertEquals(rs.getString(1), rs.getObject(1, String.class));
```
Comment (hawkfish) [inline tools/jdbc/src/test/java/org/duckdb/test/TestDuckDBJDBC.java:453]: Should we test the values as well as the types?

File: tools/jdbc/src/test/java/org/duckdb/test/TestDuckDBJDBC.java [hunk 3]
```diff
@@ -410,7 +411,63 @@ public static void duckdb_timestamp_test() throws Exception {
...
+		assertEquals(rs.getTime(9), rs.getObject(9, Time.class));
+		assertEquals(rs.getTimestamp(10), rs.getObject(10, Timestamp.class));
+		assertEquals(rs.getObject(10, LocalDateTime.class), LocalDateTime.parse("1970-01-03T03:42:23"));
+
+		rs.close();
+		ps.close();
+		stmt.close();
+		conn.close();
+	}
+
```
Comment (hawkfish) [inline tools/jdbc/src/test/java/org/duckdb/test/TestDuckDBJDBC.java:470]: Can we add failure tests for the types that aren't implemented?

**Review by hawkfish — COMMENTED** [review]
Looks good (modulo my imperfect understanding of the Java APIs!). Just some small testing comments.

File: tools/jdbc/src/test/java/org/duckdb/test/TestDuckDBJDBC.java (hunk 1 above)
Comment (Jens-H) [reply tools/jdbc/src/test/java/org/duckdb/test/TestDuckDBJDBC.java:420]: The LocalDateTime internally consists of a LocalDate and a LocalTime. And these store the date/time just as ints. So there is no dependency on a testing time zone or offset that I could think of. 
I could make this more explicit by using a different constructor like: `LocalDateTime.of(2021,1,18,21,20,7)`

References:
https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/time/LocalDateTime.java
https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/time/LocalDate.java
https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/time/LocalTime.java

File: tools/jdbc/src/test/java/org/duckdb/test/TestDuckDBJDBC.java (hunk 2 above)
Comment (Jens-H) [reply tools/jdbc/src/test/java/org/duckdb/test/TestDuckDBJDBC.java:453]: The values should be equal to pass. assertEquals(a, b) basically calls a.equals(b) which typically should compare the values of a and b. 

E.g. this is false:
`LocalDateTime.of(1999,1,1,1,1).equals(LocalDateTime.of(2000,1,1,1,1))`

File: tools/jdbc/src/test/java/org/duckdb/test/TestDuckDBJDBC.java (hunk 3 above)
Comment (Jens-H) [reply tools/jdbc/src/test/java/org/duckdb/test/TestDuckDBJDBC.java:470]: That's a fair point. Though there kind of two levels: 
1. Object types that are entirely not covered, e.g. OffsetDateTime for timestamps with tz
2. Object types that could use the database type after a cast, e.g. getObject(1, Integer.class) for a LogicalTypeId::SMALLINT

I have to see how many more tests that would be and then decide if it is feasible to do both.

File: tools/jdbc/src/test/java/org/duckdb/test/TestDuckDBJDBC.java (hunk 1 above)
Comment (hawkfish) [reply tools/jdbc/src/test/java/org/duckdb/test/TestDuckDBJDBC.java:420]: Sounds good.

File: tools/jdbc/src/test/java/org/duckdb/test/TestDuckDBJDBC.java (hunk 2 above)
Comment (hawkfish) [reply tools/jdbc/src/test/java/org/duckdb/test/TestDuckDBJDBC.java:453]: Ah, I thought it was returning types for comparison. My bad.

File: tools/jdbc/src/test/java/org/duckdb/test/TestDuckDBJDBC.java (hunk 3 above)
Comment (Jens-H) [reply tools/jdbc/src/test/java/org/duckdb/test/TestDuckDBJDBC.java:516]: According to the JDBC ref. (Table B-3) I would conclude that 1. is sufficient.

File: tools/jdbc/src/main/java/org/duckdb/DuckDBResultSet.java [hunk 4]
```diff
@@ -1214,82 +1215,117 @@ public void updateNClob(String columnLabel, Reader reader) throws SQLException {
...
 		// Compare results with expected results from Javadoc
 		// https://docs.oracle.com/en/java/javase/17/docs/api/java.sql/java/sql/ResultSet.html
-		// Also all called methods should really be implemented ... :(
 		if (type == BigDecimal.class) {
-		  if (sqlType == DuckDBColumnType.DECIMAL) {
-			return type.cast(getBigDecimal(columnIndex));
-		  } else {
-			throw new SQLException("Can't convert value to BigDecimal " + type.toString());
-		  }
+		  	if (sqlType == DuckDBColumnType.DECIMAL) {
```
Comment (hawkfish) [inline tools/jdbc/src/main/java/org/duckdb/DuckDBResultSet.java:1219]: Formatting looks strange?

**Review by hawkfish — APPROVED** [review]
Thanks!  Just have a look at what happened with the formatting?

**Comment by Mytherin** [comment]
@hawkfish can this be merged?

**Comment by hawkfish** [comment]
> @hawkfish can this be merged?

Yes.

**Comment by Mytherin** [comment]
Thanks for the PR!

**Pull request merged**